src/i2c.rs

```diff
@@ -256,202 +336,175 @@ impl AddressMode for SevenBitAddress {}
+        type Bus: I2cBusBase;
 
-    /// Blocking I2C
-    pub trait I2c<A: AddressMode = SevenBitAddress>: ErrorType {
-        /// Reads enough bytes from slave with `address` to fill `buffer`
-        ///
-        /// # I2C Events (contract)
-        ///
-        /// ``` text
-        /// Master: ST SAD+R        MAK    MAK ...    NMAK SP
-        /// Slave:           SAK B0     B1     ... BN
-        /// ```
-        ///
-        /// Where
-        ///
-        /// - `ST` = start condition
-        /// - `SAD+R` = slave address followed by bit 1 to indicate reading
-        /// - `SAK` = slave acknowledge
-        /// - `Bi` = ith byte of data
-        /// - `MAK` = master acknowledge
-        /// - `NMAK` = master no acknowledge
-        /// - `SP` = stop condition
-        fn read(&mut self, address: A, buffer: &mut [u8]) -> Result<(), Self::Error>;
-
-        /// Writes bytes to slave with address `address`
+        /// Perform a transaction against the device.
         ///
-        /// # I2C Events (contract)
+        /// - Locks the bus
+        /// - Calls `f` with an exclusive reference to the bus, which can then be used to do transfers against the device.
+        /// - Does a [stop condition](I2cBus::stop) on the bus.
+        /// - [Flushes](I2cBus::flush) the bus.
+        /// - Unlocks the bus.
         ///
-        /// ``` text
-        /// Master: ST SAD+W     B0     B1     ... BN     SP
-        /// Slave:           SAK    SAK    SAK ...    SAK
-        /// ```
-        ///
-        /// Where
-        ///
-        /// - `ST` = start condition
-        /// - `SAD+W` = slave address followed by bit 0 to indicate writing
-        /// - `SAK` = slave acknowledge
-        /// - `Bi` = ith byte of data
-        /// - `SP` = stop condition
-        fn write(&mut self, address: A, bytes: &[u8]) -> Result<(), Self::Error>;
-
-        /// Writes bytes to slave with address `address`
+        /// The locking mechanism is implementation-defined. The only requirement is that it must prevent two
+        /// transactions from executing concurrently on the same bus. Examples of implementations are:
+        /// critical sections, blocking mutexes, returning an error or panicking if the bus is already busy.
+        fn transaction<R>(
+            &mut self,
+            f: impl FnOnce(&mut Self::Bus) -> Result<R, <Self::Bus as ErrorType>::Error>,
+        ) -> Result<R, Self::Error>;
+
+        /// Do a write within a transaction.
         ///
-        /// # I2C Events (contract)
+        /// This is a convenience method equivalent to `device.transaction(|bus| { bus.start(address, Write); bus.write(buf) })`.
         ///
-        /// Same as the `write` method
-        fn write_iter<B>(&mut self, address: A, bytes: B) -> Result<(), Self::Error>
+        /// See also: [`I2cDevice::transaction`], [`I2cBusBase::write`]
+        fn write<A: AddressMode>(&mut self, address: A, buf: &[u8]) -> Result<(), Self::Error>
```

Comment:
I know this goes into a completely different direction but I want to at least have the idea discussed:  Should we tie the device address to the `I2cDevice` instead, such that the driver code never mentions it directly?

Prior art for this is the way the Linux kernel manages I2C devices.  Drivers get a `struct i2c_client*` which is already tied to a specific address.  The driver then only issues writes and reads to the `i2c_client` without ever bothering with the devices specific address.

I see a number of advantages to this:

- Device-drivers cannot accidentally interact with a wrong device on the bus - they only see the specific client that they are supposed to interact with.
- The address, which is part of the board configuration, is no longer managed by the driver and thus something like a device-tree structure can be implemented here as well.  Right now, drivers all have their own creative ways of address initialization which makes this rather difficult.
- We can have "bus" implementations that hide details behind the address like an I2C mux.  The `I2cDevice` handles the muxing opaquely so the driver still sees no difference.

The two downsides I see are

- If there ever is a driver that needs to access two devices with separate addresses, it needs to get two separate `I2cDevice` handles.  This doesn't seem to be a problem in Linux so I'm not sure how bad it would be for us.
- The initialization code will be a bit more verbose because there are now two steps (address/bus, then device).  Roughly:
  ```rust
  use i2c_encabulator::{EncabulatorAddress, EncabulatorDriver};
  
  let i2c_dev = bus.acquire(EncabulatorAddress::with_config_pins(true, false, true).into());
  let encabulator = EncabulatorDriver::new(i2c_dev);
  ```

Reply:
(And I think this would be more inline with the way we handle `SpiDevice`:  The `SpiDevice` contains the chip-select management so a driver doesn't care about the "address" on the SPI-bus)

Reply:
This is an interesting proposal. Here some additional background info from my own experience.

There definitely are devices that use several addresses. From the top of my head, for example the fairly popular [lsm303agr](https://github.com/eldruin/lsm303agr-rs) has two addresses: accelerometer at `0x19`, magnetometer at `0x1E` (the one in the newer micro:bit boards).

Another downside is that the user needs to know the address of the device to use the driver, while at the moment this is all handled by the driver. This will lead again to a fair amount of creative ways to expose the address in the driver via some public constant (which may actually be wrongly aliased on import if using multiple devices: `use mlx90614::ADDRESS as DISPLAY_ADDRESS`, this is speculative though) or whatever.

One additional downside is when using devices which have a software-programmable address. For example, the classic [MLX90614](https://github.com/eldruin/mlx9061x-rs) IR thermometer. At the moment [supporting this is trivial](https://github.com/eldruin/mlx9061x-rs/blob/3708e093420b6c068ed56d8dac6429b66a68b51d/src/common.rs#L26), just store the address as a data member and update it as necessary.
If the address is tied to the `I2cDevice` this would be slightly more complicated. For example it would need a method that takes the new address and a new `I2cDevice` with the new address (note: these have to match, unclear how to verify this, maybe we should then expose the address in the device) and replaces the device in the driver (assuming the `I2cDevice` implementation type does not change, please do not add the address as part of the type in your impl :).
Going from a 7-bit address to a 10-bit address would be more complicated since the actual `I2cDevice` type changes, but 10-bit devices are less common anyway (I3C removed 10-bit addressing completely).

There is also stuff like the general call address (`0x0` + command byte), which may be used to issue a command for all devices, for example a software reset (`0x0`+`0x6`). In that case, we would need to provide the driver with the whole bus, which actually makes sense, although it makes it slightly more complicated for the case of a single device.

Still I think this proposal is interesting and I like it from a pure-software perspective.